api/src/services/graphql/index.ts

```diff
@@ -156,6 +156,12 @@ export class GraphQLService {
 			throw new InvalidPayloadException('GraphQL execution error.', { graphqlErrors: [err.message] });
 		}
 
+		if (!result?.data && result?.errors) {
+			throw new InvalidPayloadException('GraphQL execution error.', {
+				graphqlErrors: result.errors.map((error) => error.message),
+			});
+		}
+
 		const formattedResult: FormattedExecutionResult = {
 			...result,
 			errors: result.errors?.map((error) => processError(this.accountability, error)),
```

Comment:
This solution does make perfect sense (since throwing will ensure it will not enter the `respond` middleware which caches responses)!

But would you mind sharing a way to reproduce an error that is not entering the catch block? I tried to reproduce GraphQL errors but I was only able to reach the catch block before this if statement 😅

Also out of curiosity, would this mean line 167 will never be used/reached now? if it is, should it be removed? Unless there's an actual situation where both `result.data` and `result.errors` are present of course, but I'm not sure how that can occur.

---

Apart from that, I've also noticed the usage of `res.locals.cache = false` over here to not cache GraphQL mutations:

https://github.com/directus/directus/blob/dc1a218c58110fc28ec58a43b817069ec72df173/api/src/middleware/graphql.ts#L59-L62

I'm wondering what if we use a similar approach and add something such as:

```js
if(!res.locals.payload?.data && res.locals.payload?.errors) {
    res.locals.cache = false;
}
```

to the GraphQL controller here (between these 2 lines):

https://github.com/directus/directus/blob/dc1a218c58110fc28ec58a43b817069ec72df173/api/src/controllers/graphql.ts#L36-L38

Not sure can that be a potential solution as well, and maybe retains the usage of `processError` in formattedResult? (but I wasn't able to confirm this since I couldn't replicate a non-catch graphql error)

Reply:
> This solution does make perfect sense (since throwing will ensure it will not enter the `respond` middleware which caches responses)!
> 
> But would you mind sharing a way to reproduce an error that is not entering the catch block? I tried to reproduce GraphQL errors but I was only able to reach the catch block before this if statement 😅
- By putting a `throw new Error('...')` in `readByQuery` within `services/items.ts`. I've done it like this:
```ts
	async readByQuery(query: Query, opts?: QueryOptions): Promise<Item[]> {
		if (this.collection === 'package') {
			throw new Error('Oops');
		}
```


> Also out of curiosity, would this mean line 167 will never be used/reached now? if it is, should it be removed? Unless there's an actual situation where both `result.data` and `result.errors` are present of course, but I'm not sure how that can occur.
- I am not sure 🤔 But I remember that I already had both data and errors in same Graphql response. Just don't remember if it was with Directus or not 😅


> Apart from that, I've also noticed the usage of `res.locals.cache = false` over here to not cache GraphQL mutations:
> 
> https://github.com/directus/directus/blob/dc1a218c58110fc28ec58a43b817069ec72df173/api/src/middleware/graphql.ts#L59-L62
> 
> I'm wondering what if we use a similar approach and add something such as:
> 
> ```js
> if(!res.locals.payload?.data && res.locals.payload?.errors) {
>     res.locals.cache = false;
> }
> ```
> 
> to the GraphQL controller here (between these 2 lines):
> 
> https://github.com/directus/directus/blob/dc1a218c58110fc28ec58a43b817069ec72df173/api/src/controllers/graphql.ts#L36-L38
> 
> Not sure can that be a potential solution as well, and maybe retains the usage of `processError` in formattedResult? (but I wasn't able to confirm this since I couldn't replicate a non-catch graphql error)
- That seems a good solution, but the response status will always be 200 which does not seem right since response has errors

Reply:
> - By putting a `throw new Error('...')` in `readByQuery` within `services/items.ts`. 

Thanks! Works like a charm 😁

> - I am not sure 🤔 But I remember that I already had both data and errors in same Graphql response. Just don't remember if it was with Directus or not 😅

It seems like before this PR, and using the `throw new Error()` in `readByQuery` above, the response does show both since `...result` is spread:

Note: this is using the public role and in production (or NODE_ENV not `'development'`), because error handler middleware shows stack trace when `NODE_ENV` is 'development'

<table>
<tr>
<td> Before </td> <td> After </td>
</tr>
<tr>
<td>

```js
{
  "errors": [
    {
      "message": "An unexpected error occurred.",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": null
}
```

</td>
<td>
    
```js
{
  "errors": [
    {
      "message": "GraphQL execution error.",
      "extensions": {
        "code": "INVALID_PAYLOAD",
        "graphqlErrors": [
          "Oops"
        ]
      }
    }
  ]
}
```

</td>
</tr>
</table>

I'm not entirely sure are there any need to change the current error response, so I'll have to defer to @rijkvanzanten, and possibly @licitdev for the work in #16430 which I believe was also meant to address #16312. Maybe we should process the error as Ian did in #16430, since this PR does still surface the sql error when using a non-existent sort field:

<table>
<tr>
<td> Before </td> <td> After </td>
</tr>
<tr>
<td>

```js
{
  "errors": [
    {
      "message": "An unexpected error occurred.",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": null
}
```

</td>
<td>
    
```js
{
  "errors": [
    {
      "message": "GraphQL execution error.",
      "extensions": {
        "code": "INVALID_PAYLOAD",
        "graphqlErrors": [
          "select \"collection_a\".\"name\", \"collection_a\".\"id\" from \"collection_a\" order by \"collection_a\".\"test\" asc limit $1 - column collection_a.test does not exist"
        ]
      }
    }
  ]
}
```

</td>
</tr>
</table>

I've also noticed based on the graphql spec: https://spec.graphql.org/October2021/#sec-Data, there's a slight distinction between "data" not being present in the response, vs data responded as null. I wasn't aware of this before, but mentioning it here just in case the original null data was meant to adhere to this (but doesn't seem like it 100% is).

> That seems a good solution, but the response status will always be 200 which does not seem right since response has errors 

Totally agree that it doesn't seem right to respond with 200 when there's error, but I also vaguely recall @licitdev or @rijkvanzanten discussing about this, where GraphQL typically still returns 200 status code even for errors. As for whether it should return `4xx`/`5xx` vs `200` for errors, I've found many contradicting answers so I'll once again defer to others on what should the API respond with.

I was able to find this GraphQL over HTTP spec that is still in draft, but it does interestingly state the following:

> If the response uses a non-`200` status code and the media type of the response payload is `application/json` then the client MUST NOT rely on the body to be a well-formed GraphQL response since the source of the response may not be the server but instead some intermediary such as API gateways, proxies, firewalls, etc.

(link to above quote: https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#processing-the-response)

and possibly only response body with `application/graphql-response+json` media type will be allowed to use `4xx`/`5xx` 🤔

Reply:
> Unless there's an actual situation where both result.data and result.errors are present of course, but I'm not sure how that can occur.

There is! You can query multiple top level collections at the same time, and one of them can error out, while the rest succeeds.

> I'm wondering what if we use a similar approach and add something such as:

I'd prefer this approach! It's way more explicit and doesn't circumvent the respond handler, which is also in charge of exporting and some other things.

> but the response status will always be 200 which does not seem right since response has errors

I know right! However, this is per the GraphQL spec! GraphQL doesn't use HTTP status codes to communicate status (unlike REST), so 200 for everything is correct..